design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+
+We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.
+
+Thus, we will not use this option due to security concerns.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
+
+In this case, we respond with 202 Accepted to the "inject-client-cert" request and start injecting the cert in a separate thread. The server then polls on the state of the authn-client container and waits until its State changes to some  agreed State (e.g "CertInjected"). This state will be set by the authn-client once it sees that the cert is installed.
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
+#### Option 4 - Send cert as part of bash script
+
+The issue with the STDIN that is not closing is an issue of the `tar` command that waits for more input and doesn't close the connection. Another option that we have is to change the command from `tar` to `sh` and send a bash script in the STDIN.
+The script can be something like:
+
+```
+#!/bin/sh
+
+cat > /etc/conjur/ssl/client.pem <<EOF
+#{cert-content}
+EOF
+
+chmod 644 /etc/conjur/ssl/client.pem
+exit
+```
+
+This approach has the following benefits:
+1. Short URL. Certificate is in the script. (avoid issues raised from option 2)
+1. Deterministic operation. You know exactly when it is done, because the connection is closed once it reaches exit.
+1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
+1. Versatility. Script can be customized later if needed to handle more issues in this area.
+1. Simple implementation
+
+This solution is implemented in [this PR](https://github.com/cyberark/conjur/pull/1789).
+
+#### Conclusion
+
+Option 2 is off the table due to security concerns.
+
+Option 3 is off the table due to redundant complexity and backwards-compatibiliy issues.
+
+Option 4 has the most benefits of the options and is better than option 1.
+
+<u>Decision</u>: Option 4
+
+### Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+The authentication process is depicted in the following diagram:
+![Authn-k8s authenti](authn-k8s-process.png)
```

Comment:
in the process diagram, it seems to show that the server validates the request, kicks off the cert injection, and closes the socket before sending the 202 response. this isn't clear from the text of the doc, however - can a note about this be added?

also, is it possible to use plantuml and include the source for this process diagram in this PR? we could perhaps create a `design/diagrams` folder in this project, and put both the .txt source and the .png diagram in that folder. for an example, please see the [Puppet diagrams folder](https://github.com/cyberark/conjur-puppet/tree/master/diagrams)

Reply:
this is described in [this section](https://github.com/cyberark/conjur/blob/kubectl-exec-solution-design/design/authenticators/authn_k8s/inject_client_cert_solution_design.md#waiting-for-the-injection-to-finish).

The source for the diagram is not accesible as it's done in Lucidchart under my license. i'll consider using `plantuml` for this.